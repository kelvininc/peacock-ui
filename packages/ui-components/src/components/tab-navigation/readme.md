# kv-tab-navigation



<!-- Auto Generated Below -->


## Usage

### Angular

```html
<!-- With required props -->
<kv-tab-navigation [tabs]="myNavigationTabs" selectedTabKey="tab1"></kv-tab-navigation>

<!-- With required props and notification -->
<kv-tab-navigation [tabs]="myNavigationTabs" selectedTabKey="tab1" [notifications]="myNavigationTabsNotifications"></kv-tab-navigation>
```


### React

```tsx
import React from 'react';
import { KvTabNavigation } from '@kelvininc/react-ui-components';

export const TabNavigationExample: React.FC = () => (
	<>
		{/*-- With required props --*/}
		<KvTabNavigation tabs={myNavigationTabs} selectedTabKey="tab1"></KvTabNavigation>

		{/*-- With required props and notification --*/}
		<KvTabNavigation
			tabs={myNavigationTabs}
			selectedTabKey="tab1"
			notifications={myNavigationTabsNotifications}>
		</KvTabNavigation>
	</>
);
```



## Properties

| Property            | Attribute          | Description                                                                     | Type                                           | Default                |
| ------------------- | ------------------ | ------------------------------------------------------------------------------- | ---------------------------------------------- | ---------------------- |
| `notifications`     | --                 | (optional) To add a notification dot and its respective color to a specific tab | `ITabsNotificationDict`                        | `{}`                   |
| `selectedTabKey`    | `selected-tab-key` | (optional) The currently selected tab key                                       | `number \| string`                             | `undefined`            |
| `size`              | `size`             | (optional) Sets the items on this tab nav to use small styling configuration    | `EComponentSize.Large \| EComponentSize.Small` | `EComponentSize.Large` |
| `tabs` _(required)_ | --                 | (required) The tab items to render in this component                            | `ITabNavigationItem[]`                         | `undefined`            |


## Events

| Event       | Description                                                  | Type                  |
| ----------- | ------------------------------------------------------------ | --------------------- |
| `tabChange` | When the tab selection changes, emit the requested tab's key | `CustomEvent<string>` |


## Dependencies

### Depends on

- [kv-tab-list](../tab-list)
- [kv-tab-item](../tab-item)

### Graph
```mermaid
graph TD;
  kv-tab-navigation --> kv-tab-list
  kv-tab-navigation --> kv-tab-item
  style kv-tab-navigation fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


