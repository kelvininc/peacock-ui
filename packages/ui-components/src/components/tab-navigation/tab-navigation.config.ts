export const INTERSECTION_OBSERVER_CONFIG: IntersectionObserverInit = {
	root: null,
	rootMargin: '0px',
	threshold: 1.0
};

export const DEFAULT_INDICATOR_TIMEOUT_WAIT = 0;
