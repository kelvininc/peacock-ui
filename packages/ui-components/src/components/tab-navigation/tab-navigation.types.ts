export interface ITabNavigationItem {
	tabKey: number | string;
	label: string;
	disabled?: boolean;
}

export interface ITabNotification {
	active: boolean;
	color?: string;
}

export interface ITabsNotificationDict {
	[tabKey: string]: ITabNotification;
	[tabKey: number]: ITabNotification;
}

export type ISelectedTabIndicatorConfig = {
	left: string;
	width: string;
};
