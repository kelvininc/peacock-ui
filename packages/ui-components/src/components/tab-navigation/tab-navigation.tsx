import { Component, Element, Event, EventEmitter, Host, Listen, Prop, State, Watch, h } from '@stencil/core';
import { ISelectedTabIndicatorConfig, ITabNavigationItem, ITabsNotificationDict } from './tab-navigation.types';

import { EComponentSize } from '../../utils/types';
import { findTabElement, getRelativeClientRect } from './tab-navigation.utils';
import { HTMLStencilElement } from '@stencil/core/internal';

@Component({
	tag: 'kv-tab-navigation',
	styleUrl: 'tab-navigation.scss',
	shadow: true
})
export class KvTabNavigation {
	/** (required) The tab items to render in this component */
	@Prop({ reflect: true }) tabs!: ITabNavigationItem[];
	/** (optional) The currently selected tab key */
	@Prop() selectedTabKey?: number | string;
	/** (optional) To add a notification dot and its respective color to a specific tab */
	@Prop() notifications?: ITabsNotificationDict = {};
	/** (optional) Sets the items on this tab nav to use small styling configuration */
	@Prop() size?: EComponentSize = EComponentSize.Large;

	/** When the tab selection changes, emit the requested tab's key */
	@Event() tabChange: EventEmitter<string>;

	/** Listen to custom DOM event of tab selection */
	@Listen('tabSelected')
	tabSelectionHandler(event: CustomEvent<string>) {
		this.tabChange.emit(event.detail);
	}

	/** Watch for tab selection change and react accordingly by updating the internal states */
	@Watch('selectedTabKey')
	tabSelectionChangeHandler() {
		this.calculateTabIndicatorPosition();
	}

	/** The left offset and width of the tab indicator, recalculated when the selected tab changes */
	@State() selectedTabIndicatorConfig: ISelectedTabIndicatorConfig = {
		left: '0px',
		width: '0px'
	};

	@Element() el: HTMLKvTabNavigationElement;

	initialRenderIntervalId: number;

	componentDidLoad() {
		const selectedTabEl = findTabElement(this.el, this.tabs, this.selectedTabKey);
		this.setIntervalUntilElementIsVisible(selectedTabEl);
	}

	disconnectedCallback() {
		window.clearInterval(this.initialRenderIntervalId);
	}

	private setIntervalUntilElementIsVisible = (element: HTMLStencilElement, intervalMs: number = 100): void => {
		this.initialRenderIntervalId = window.setInterval(() => {
			const rect = getRelativeClientRect(element);

			if (rect && rect.width > 0) {
				window.clearInterval(this.initialRenderIntervalId);
				this.updateTabIndicatorConfig(rect);
				return;
			}
		}, intervalMs);
	};

	private updateTabIndicatorConfig = (rect: DOMRect) => {
		this.selectedTabIndicatorConfig = {
			left: `${rect.left}px`,
			width: `${rect.width}px`
		};
	};

	private calculateTabIndicatorPosition() {
		const selectedTabEl = findTabElement(this.el, this.tabs, this.selectedTabKey);

		if (selectedTabEl) {
			this.updateTabIndicatorConfig(getRelativeClientRect(selectedTabEl));
		}
	}

	render() {
		return (
			<Host>
				{this.tabs.map(item => (
					<kv-tab-item
						tabKey={item.tabKey}
						label={item.label}
						disabled={item.disabled}
						selected={item.tabKey === this.selectedTabKey}
						size={this.size}
						hasNotification={this.notifications[item.tabKey]?.active}
						notificationColor={this.notifications[item.tabKey]?.color}
					></kv-tab-item>
				))}
				<div class="selected-tab-indicator" style={this.selectedTabIndicatorConfig}></div>
			</Host>
		);
	}
}
