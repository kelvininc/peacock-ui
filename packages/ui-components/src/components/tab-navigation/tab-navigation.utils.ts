import { EComponentSize } from '../../types';
import { FONT_DESCRIPTOR_SIZE_MAP, TAB_ITEM_NOTIFICATION_DOT_WITH_MARGIN, TAB_ITEM_PADDING_PX, TAB_ITEM_SMALL_ADDED_MARGIN } from './tab-navigation.config';
import { ISelectedTabIndicatorConfig, ITabNavigationItem, ITabsNotificationDict } from './tab-navigation.types';

export const calculateTabWidths = (tabs: ITabNavigationItem[], notifications: ITabsNotificationDict, size: EComponentSize) => {
	const measuringCanvas = document.createElement('canvas');
	const font = FONT_DESCRIPTOR_SIZE_MAP[size];
	const ctx = measuringCanvas.getContext('2d');
	ctx.font = font;

	const values = tabs.reduce<Record<string, ISelectedTabIndicatorConfig>>((acc, { tabKey, label }, idx) => {
		const previousTab = idx > 0 ? tabs[idx - 1] : undefined;
		const leftOffset = previousTab ? +acc[previousTab.tabKey].width + +acc[previousTab.tabKey].left : 0;
		const notificationDotWidth = notifications[tabKey]?.active ? TAB_ITEM_NOTIFICATION_DOT_WITH_MARGIN : 0;
		const small = size === EComponentSize.Small ? TAB_ITEM_SMALL_ADDED_MARGIN : 0;
		const textWidth = Math.ceil(ctx.measureText(label).width);

		acc[tabKey] = {
			left: `${leftOffset}`,
			width: `${textWidth + TAB_ITEM_PADDING_PX + notificationDotWidth + small}`
		};

		return acc;
	}, {});

	return values;
};
