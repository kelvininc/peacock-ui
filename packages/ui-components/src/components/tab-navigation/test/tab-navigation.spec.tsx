import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { cloneDeep } from 'lodash-es';
import { KvTabNavigation } from '../tab-navigation';
import { NOTIFICATIONS_MOCK, TAB_ITEMS } from './tab-navigation.mock';
import { h } from '@stencil/core';

describe('Tab Navigation (unit tests)', () => {
	let page: SpecPage;
	let component: KvTabNavigation;

	const tabsMock = cloneDeep(TAB_ITEMS);
	const notificationsMock = cloneDeep(NOTIFICATIONS_MOCK);

	describe('when rendering with required props', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvTabNavigation],
				template: () => <kv-tab-navigation selectedTabKey="tab1" tabs={tabsMock}></kv-tab-navigation>
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		it('should set the correct component tabs', () => {
			expect(component.tabs).toEqual([
				{
					tabKey: 'tab1',
					label: 'Tab 1'
				},
				{
					tabKey: 'tab2',
					label: 'Tab 1'
				},
				{
					tabKey: 'tab3',
					label: 'Tab 1'
				},
				{
					tabKey: 'tab4',
					label: 'Tab 1'
				},
				{
					tabKey: 'tab5',
					label: 'Tab 1'
				}
			]);
		});
	});

	describe('when rendering with required props and notifications', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvTabNavigation],
				template: () => <kv-tab-navigation selectedTabKey="tab1" tabs={tabsMock} notifications={notificationsMock}></kv-tab-navigation>
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		it('should set the correct notifications', () => {
			expect(component.notifications).toEqual({
				tab1: {
					active: true
				}
			});
		});
	});
});
