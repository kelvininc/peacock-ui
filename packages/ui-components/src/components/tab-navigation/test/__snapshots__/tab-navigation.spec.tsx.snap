// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab Navigation (unit tests) when rendering with required props and notifications should match the snapshot 1`] = `
<kv-tab-navigation>
  <mock:shadow-root>
    <kv-tab-item hasnotification="" label="Tab 1" selected="" size="large" tabkey="tab1"></kv-tab-item>
    <kv-tab-item label="Tab 1" size="large" tabkey="tab2"></kv-tab-item>
    <kv-tab-item label="Tab 1" size="large" tabkey="tab3"></kv-tab-item>
    <kv-tab-item label="Tab 1" size="large" tabkey="tab4"></kv-tab-item>
    <kv-tab-item label="Tab 1" size="large" tabkey="tab5"></kv-tab-item>
    <div class="selected-tab-indicator" style="left: undefinedpx; width: undefinedpx;"></div>
  </mock:shadow-root>
</kv-tab-navigation>
`;

exports[`Tab Navigation (unit tests) when rendering with required props should match the snapshot 1`] = `
<kv-tab-navigation>
  <mock:shadow-root>
    <kv-tab-item label="Tab 1" selected="" size="large" tabkey="tab1"></kv-tab-item>
    <kv-tab-item label="Tab 1" size="large" tabkey="tab2"></kv-tab-item>
    <kv-tab-item label="Tab 1" size="large" tabkey="tab3"></kv-tab-item>
    <kv-tab-item label="Tab 1" size="large" tabkey="tab4"></kv-tab-item>
    <kv-tab-item label="Tab 1" size="large" tabkey="tab5"></kv-tab-item>
    <div class="selected-tab-indicator" style="left: undefinedpx; width: undefinedpx;"></div>
  </mock:shadow-root>
</kv-tab-navigation>
`;
